### PR TITLE
feat: NeuMF

### LINKED ISSUE #126: [FEAT] replace TruncatedSVD with NeuMF
## Feature Description
Replace the current `TruncatedSVD`-based collaborative filtering in `collaborative_model.py` with a Neural Matrix Factorization (NeuMF) model — a two-tower ANN architecture combining a Generalized Matrix Factorization (GMF) branch and a Multi-Layer Perceptron (MLP) branch, trained end-to-end on implicit feedback (clicks/purchases).

## Problem it Solves
TruncatedSVD is a linear model — it represents user-item affinity as a simple dot product of latent vectors. This means it can only capture linear relationships and completely misses complex, non-linear preference patterns. For example:
- A user who buys hiking boots and waterproof jackets probably wants trekking poles — but this three-way co-occurrence pattern is invisible to a dot product.
- SVD has no mechanism to weight implicit feedback signals differently (a purchase vs a click carry very different intent, but SVD treats them the same).
- Latent vectors learned by SVD are not trained end-to-end — they are decomposed offline and cannot adapt to auxiliary signals.

NeuMF adds a non-linear interaction layer (the MLP branch) on top of learnable embeddings, trained end-to-end with binary cross-entropy on implicit feedback, which directly addresses all three limitations.

## Proposed Solution
Implement NeuMF in `collaborative_model.py` using PyTorch:

**Architecture:**
```
User ID ──→ GMF Embedding ──┐
                             ├──→ Element-wise product ──→ GMF out ──┐
Item ID ──→ GMF Embedding ──┘                                         ├──→ Concat ──→ Linear ──→ Sigmoid
                                                                       │
User ID ──→ MLP Embedding ──┐                                         │
                             ├──→ Concat ──→ Dense+ReLU layers ──→ MLP out ──┘
Item ID ──→ MLP Embedding ──┘
```

**Implementation steps:**
1. Add `torch` and `torch.nn` to `requirements.txt`
2. In `collaborative_model.py`, implement `NeuMF(nn.Module)` with:
   - Separate GMF and MLP embedding tables (`nn.Embedding`)
   - GMF branch: element-wise product of user + item GMF embeddings
   - MLP branch: concatenated user + item MLP embeddings → 3 Dense + ReLU + Dropout layers (e.g. 256→128→64)
   - Fusion: concatenate GMF out + MLP out → single `nn.Linear` → sigmoid
3. Training loop: binary cross-entropy loss, Adam optimizer, negative sampling (sample uninteracted items as negatives)
4. Keep `TruncatedSVD` as a fallback behind a `USE_NEUMF=True` flag in `.env` so existing deployments are not broken
5. Expose the same `.get_recommendations(user_id, top_k)` interface so `hybrid_model.py` requires zero changes
6. Add benchmark in `evaluation.py` comparing Precision@K, Recall@K, NDCG@K between SVD and NeuMF

**Embedding dimensions:** 64 for GMF, 64 for MLP (configurable via `.env`)
**Training:** ~20 epochs, batch size 256, runs on CPU in under 2 minutes on the existing dataset size

## Alternatives Considered
- **ALS (Alternating Least Squares)**: Better than SVD for implicit feedback but still linear. Good intermediate step, doesn't solve the non-linearity problem.
- **ConvNCF**: Extension of NeuMF using CNN over the outer product of embeddings (He et al. 2018). More powerful but heavier — better as a follow-up PR after NeuMF is merged.
- **LightGCN**: Graph-based collaborative filtering, state of the art but requires building a user-item interaction graph — too large a structural change for a first DL PR.

## Additional Context
- Target file: `collaborative_model.py`
- Related files: `requirements.txt`, `hybrid_model.py`, `evaluation.py`, `.env.example`
- Reference paper: He et al. 2017 — "Neural Collaborative Filtering" https://arxiv.org/abs/1708.05031
- The hybrid score formula (`α · content_score + β · collab_score + γ · sentiment_score`) stays completely unchanged — only the source of `collab_score` changes, so `hybrid_model.py` needs no edits
- PyTorch is the only new dependency added

### PR DESCRIPTION
**Pull request by parinaB**
## What changed

Replaced the existing TruncatedSVD-based collaborative filtering with a Neural Matrix Factorization (NeuMF) model implemented in PyTorch.
Added a fallback mechanism to keep SVD as default when USE_NEUMF=false, ensuring full backward compatibility.

Updated evaluation pipeline to support both SVD and NeuMF seamlessly.

## Why
SVD is a linear matrix factorization method and cannot capture complex, non-linear user–item relationships or implicit feedback patterns.

NeuMF improves this by:

- Learning non-linear interactions via MLP
- Combining GMF + MLP towers
- Supporting end-to-end training with binary cross entropy
- Using negative sampling for implicit feedback

This leads to better recommendation quality and more expressive user modeling.

## How to test

```bash
# Example:
pip install -r requirements.txt
python evaluation.py
# (Optional) Enable NeuMF
export USE_NEUMF=true
python evaluation.py
```

## Checklist
- [✓] I have read the [CONTRIBUTING.md](CONTRIBUTING.md)
- [✓] My code follows PEP8 style (`flake8 .`)
- [✓] I have tested my changes locally
- [✓] I have added/updated tests where applicable
- [✓] I can explain every line of code I've written
- [✓] I have NOT used AI-generated code without understanding and attributing it

## Related issue
Closes #126 

## AI assistance disclosure

- [ ] I did not use AI assistance for this PR
- [✓] I used AI assistance for: evaluation pipeline fixes and PR structuring

### CODE REVIEW

**Comment by parinaB** [comment]
@leonagoel could u kindly review...

**Review by leonagoel — APPROVED** [review]

**Comment by leonagoel** [comment]
✅ Marked as approved but **cannot auto-merge** — this PR has conflicts.

Please resolve conflicts and push again, then re-add the `status:approved` label.

**Review by leonagoel — CHANGES_REQUESTED** [review]
Thanks for the PR, @parinaB! NeuMF (Neural Matrix Factorization) is a valuable addition to the recommender system. However there are **conflicts in 7 files** that need to be resolved first.

**Conflicting files:**
- `collaborative_model.py`
- `content_model.py`
- `data_adapter.py`
- `data_preprocessing.py`
- `dataset_manager.py`
- `evaluation.py`
- `hybrid_model.py`

This looks like your branch may be significantly behind `main`. Please do a full rebase:
```
git fetch origin
git rebase origin/main
# resolve conflicts file by file
git add collaborative_model.py content_model.py data_adapter.py data_preprocessing.py dataset_manager.py evaluation.py hybrid_model.py
git rebase --continue
git push --force-with-lease
```

Since there are many conflicts, you may find it easier to click **"Resolve conflicts"** on GitHub for each file individually.

Once resolved and CI passes, this will be auto-merged. Please address within **48 hours**.

**Comment by parinaB** [comment]
@leonagoel resolved.. please take a look .. thank you

**Comment by parinaB** [comment]
@leonagoel i have made the necessary resolves.

**Comment by parinaB** [comment]
@leonagoel any updates??

**Comment by parinaB** [comment]
updates??
